hydrogen-music/hydrogen
# Velocity automation

# Overview

This branch allows users to adjust velocity of played notes via a timeline called "automation path":
![screenshot](https://dl.dropboxusercontent.com/u/1474460/Hydrogen/automation_paths.png)
[audio](https://soundcloud.com/el-pescado/ap-demo/s-Vyy16)

That path allows to create more complex songs using fewer patterns. It works especially nice with multi-layered drumkits.
# Features
- Support for built in playback (sampler)
- Support for MIDI export
- Saving and loading songs
- Full undo&redo support
- Unit tests
# Prerequisites

This branch makes conserative use of C++11 features so it needs C++ support enabled (see pull request #263).
# Regressions

This branch should be backwards compatible. Songs created before will get constant velocity adjustment=1.0, which means note velocities will remain unchanged.

# How to test

* Check whether it's possible to add points to velocity automation path, move and delete them
* Check whether velocity is adjusted in playback in Song mode
* Check whether notes are played with their nominal velocity in Pattern mode
* Check whether note velocity is adjusted according to velocity path in exported audio file
* Check whether note velocity is adjusted according to velocity path in exported MIDI file
* Check whether after saving song with velocity automation path it's possible to load that song file, and velocity automation is correctly loaded
